### PR TITLE
ECS

### PR DESCRIPTION
**Pull request by MaelTeyssedre**

### CODE REVIEW

File: ECS/Entity/Entity.hpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
+/**
+ * \file Entity.hpp
+ *
+ * \brief file when the class Entity is defined
```
Comment (Jsavang) [inline ECS/Entity/Entity.hpp:4]: It's where, not when

**Review by MaelTeyssedre — COMMENTED** [review]
lgtm

File: ECS/Entity/Entity.hpp (hunk 1 above)
Comment (MaelTeyssedre) [reply ECS/Entity/Entity.hpp]: solved

File: ECS/Registry/Registry.hpp [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+             */
+            template <typename Component>
+            typename SparseArray<Component>::reference_type addComponent(Entity const &to, Component &&c) {
+                return (std::any_cast<SparseArray<Component> &>(_componentsArrays[std::type_index(typeid(Component))])).insertAt(to, c);
+            }
+
+            /**
+             * \fn template <typename Component, typename ...Params> typename SparseArray<Component>::reference_type emplaceComponent(Entity const &to, Params &&...p)
+             * 
+             * \brief create a nomponent for an specified entity
```
Comment (Jsavang) [inline ECS/Registry/Registry.hpp:163]: nomponent ?

File: ECS/Registry/Registry.hpp [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+            template <typename Component>
+            void removeComponent(Entity const &from) {
+                auto array = std::any_cast<SparseArray<Component> &>(_componentsArrays[std::type_index(typeid(Component))]);
+                array.erase(from);
+            }
+
+            /**
+             * \fn template <typename Function, class ...Components> void addSystem(Function &&f, Components &&...components)
+             * 
+             * \brief add a system the the registry
```
Comment (Jsavang) [inline ECS/Registry/Registry.hpp:196]: the the

File: ECS/Registry/Registry.hpp [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+            void addSystem(Function &&f, Components &&...components) {
+                _systems.push_back([&f, &components...](Registry &r) -> void {
+                    f(r, components...);
+                });
+            }
+
+            /**
+             * \fn template <typename Function, class ...Components> void addSystem(Function const &f, Components &...components)
+             * 
+             * \brief add a system the the registry
```
Comment (Jsavang) [inline ECS/Registry/Registry.hpp:214]: the the

File: ECS/Registry/Registry.hpp [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+                array.erase(from);
+            }
+
+            /**
+             * \fn template <typename Function, class ...Components> void addSystem(Function &&f, Components &&...components)
+             * 
+             * \brief add a system the the registry
+             * 
+             * \tparam Function type of function we want to move into the registry
+             * \tparam Components pack of components we want to pass in argument to the the previously passed function
```
Comment (Jsavang) [inline ECS/Registry/Registry.hpp:199]: the the

File: ECS/Registry/Registry.hpp [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+                });
+            }
+
+            /**
+             * \fn template <typename Function, class ...Components> void addSystem(Function const &f, Components &...components)
+             * 
+             * \brief add a system the the registry
+             * 
+             * \tparam Function type of function we want to pass into the registry
+             * \tparam Components pack of components we want to pass in argument to the the previously passed function
```
Comment (Jsavang) [inline ECS/Registry/Registry.hpp:217]: the the

File: ECS/Entity/Entity.hpp [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+ * \file Entity.hpp
+ *
+ * \brief file where the class Entity is defined
+ */
+
+#ifndef ENTITY_HPP_
+    #define ENTITY_HPP_
+
+    /**
+     * @brief declaration of class registry to avoid circular includes...
```
Comment (Alban-Bodenes) [inline ECS/Entity/Entity.hpp:11]: Java doc

File: ECS/Registry/Registry.hpp [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+            template <typename Function, class ...Components>
+            void addSystem(Function const &f, Components &...components) {
+                _systems.push_back([&f, &components...](Registry &r) -> void {
+                    f(r, components...);
+                });
+            }
+
+        private:
+            std::map<std::type_index, std::function<void(Registry &, Entity const &)>> _constructorArray; /*! map of default constructor for components, indexed by type */
+            std::map<std::type_index, std::function<void(Registry &, Entity const &)>> _destructorArray; /*! map of default desctructor for components, indexed by type */
```
Comment (Jsavang) [inline ECS/Registry/Registry.hpp:231]: desctructor

File: ECS/SparseArray/SparseArray.hpp [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+    #include <vector>
+    #include <iostream>
+    #include <memory>
+    #include <optional>
+    #include <type_traits>
+
+    /**
+     * \class SparseArray SparseArray.hpp
+     * 
+     * \brief teplated class for stocking components
```
Comment (Alban-Bodenes) [inline ECS/SparseArray/SparseArray.hpp:19]: teplated?

File: ECS/Registry/Registry.hpp [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+            std::map<std::type_index, std::any> _componentsArrays; /*! map of components, indexed by type */
+            std::vector<std::function<void(Registry &)>> _systems; /*! list of systems that will be executed from the first to the last (the order can be importanrt) */
```
Comment (Jsavang) [inline ECS/Registry/Registry.hpp:233]: importanrt

File: ECS/SparseArray/SparseArray.hpp [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+                for (auto i = 0; i < size; i++)
+                    _data.push_back(std::nullopt);
+            }
+
+            /**
+             * \fn SparseArray &operator=(SparseArray const &sparseArray)
+             * 
+             * \brief copy assignation operator
+             *
+             * \param sparseArray Contante reference to the SparseArray that will be assigned to another
```
Comment (Jsavang) [inline ECS/SparseArray/SparseArray.hpp:103]: Contante

File: ECS/SparseArray/SparseArray.hpp [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+                std::swap(_data, sparseArray._data);
+                return *this;
+            }
+
+            /**
+             * \fn SparseArray &operator=(SparseArray &&sparceArray) noexcept
+             * 
+             * \brief move assignation operator
+             *
+             * \param sparseArray Contante reference to the SparseArray that will be moved into another
```
Comment (Jsavang) [inline ECS/SparseArray/SparseArray.hpp:117]: Contante

File: ECS/SparseArray/SparseArray.hpp [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+             */
+            SparseArray &operator=(SparseArray &&sparceArray) noexcept {
+                _data = std::move(sparseArray._data);
+                return *this;
+            }
+
+            /**
+             * \fn reference_type operator[](size_t idx)
+             * 
+             * \brief ovderload of operator[] ti get the component at the index passed inside of bracket
```
Comment (Alban-Bodenes) [inline ECS/SparseArray/SparseArray.hpp:129]: ovderload?

File: ECS/components/components.hpp [hunk 14]
```diff
@@ -0,0 +1,73 @@
+/**
+ * \file components.hpp
```
Comment (Jsavang) [inline ECS/components/components.hpp:2]: Components.hpp may be better

File: ECS/SparseArray/SparseArray.hpp [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+                auto alloc = _data.get_allocator();
+                std::allocator_traits<decltype(alloc)>::destroy(alloc, std::addressof(_data[pos]));
+            }
+
+            /**
+             * \fn size_type getIndex(value_type const &value) const
+             * 
+             * \brief getter for the index of an entity
+             * 
+             * \param value constant reference to an antity by his value
```
Comment (Alban-Bodenes) [inline ECS/SparseArray/SparseArray.hpp:316]: Antity = Antic entity?

File: ECS/components/components.hpp [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+         */
+        struct position_s {
+            int x;
+            int y;
+        };
+
+        /**
+         * \struct position_s components.hpp
+         * 
+         * \brief velocoty component
```
Comment (Jsavang) [inline ECS/components/components.hpp:36]: velocoty

File: ECS/components/components.hpp [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+         *
+         * \param x x velocity of the element
+         * \param y y velocoty of the element
```
Comment (Jsavang) [inline ECS/components/components.hpp:39]: velocoty

File: ECS/systems/systems.hpp [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+#include "components.hpp"
+
+    /**
+     * \fn void loggingSystem(Registry &r, SparseArray<component::position_s> &positions, SparseArray<component::velocity_s> &velocities)
+     * 
+     * \brief System for logging in real-time the values of positions and velocities components
+      *
+     * \param r Reference to the registry
+     * \param positions Reference to an array of position component
+     * \param velocyties Reference to an array of velocity component
```
Comment (Jsavang) [inline ECS/systems/systems.hpp:20]: velocyties?

File: ECS/systems/systems.hpp [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+    void loggingSystem(Registry &r, SparseArray<component::position_s> &positions, SparseArray<component::velocity_s> &velocities);
+
+    /**
+     * \fn void controlSystem(Registry &r, SparseArray<component::position_s> &positions, SparseArray<component::velocity_s> &velocities, SparseArray<component::controllable_s> &controllable)
+     * 
+     * \brief System that handle update on the keyboard and update the related position component
+     *
+     * \param r Reference to the registry
+     * \param positions Reference to an array of position component
+     * \param velocoties Reference to an array of velocity component
```
Comment (Jsavang) [inline ECS/systems/systems.hpp:31]: velocoties

File: ECS/systems/systems.hpp [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+     *
+     * \param r Reference to the registry
+     * \param drawables Reference to an array of drawable component
+     */
+    void drawSystem(Registry &r, SparseArray<component::drawable_s> &drawables);
+
+    /**
+     * \fn void positionSystem(Registry &r, SparseArray<component::position_s> &positions, SparseArray<component::drawable_s> &sprite)
+     * 
+     * \brief System that update the position of the graphic element related to his coordonates
```
Comment (Jsavang) [inline ECS/systems/systems.hpp:49]: coordonate

File: ECS/systems/systems.hpp [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+
+#include "Registry.hpp"
+#include "components.hpp"
+
+    /**
+     * \fn void loggingSystem(Registry &r, SparseArray<component::position_s> &positions, SparseArray<component::velocity_s> &velocities)
+     * 
+     * \brief System for logging in real-time the values of positions and velocities components
+      *
+     * \param r Reference to the registry
```
Comment (Alban-Bodenes) [inline ECS/systems/systems.hpp]: Indentation

**Review by Alban-Bodenes — APPROVED** [review]

**Review by Jsavang — APPROVED** [review]
ltgm

File: ECS/Registry/Registry.hpp (hunk 2 above)
Comment (MaelTeyssedre) [reply ECS/Registry/Registry.hpp:163]: solved

File: ECS/Registry/Registry.hpp (hunk 3 above)
Comment (MaelTeyssedre) [reply ECS/Registry/Registry.hpp:196]: solved

File: ECS/Registry/Registry.hpp (hunk 4 above)
Comment (MaelTeyssedre) [reply ECS/Registry/Registry.hpp:214]: solved

File: ECS/Registry/Registry.hpp (hunk 5 above)
Comment (MaelTeyssedre) [reply ECS/Registry/Registry.hpp:199]: solved

File: ECS/Registry/Registry.hpp (hunk 6 above)
Comment (MaelTeyssedre) [reply ECS/Registry/Registry.hpp:217]: solved

File: ECS/Registry/Registry.hpp (hunk 8 above)
Comment (MaelTeyssedre) [reply ECS/Registry/Registry.hpp:231]: solved

File: ECS/Registry/Registry.hpp (hunk 10 above)
Comment (MaelTeyssedre) [reply ECS/Registry/Registry.hpp:233]: solved

File: ECS/SparseArray/SparseArray.hpp (hunk 11 above)
Comment (MaelTeyssedre) [reply ECS/SparseArray/SparseArray.hpp:103]: solved

File: ECS/SparseArray/SparseArray.hpp (hunk 12 above)
Comment (MaelTeyssedre) [reply ECS/SparseArray/SparseArray.hpp:117]: solved

File: ECS/components/components.hpp (hunk 16 above)
Comment (MaelTeyssedre) [reply ECS/components/components.hpp:36]: solved

File: ECS/systems/systems.hpp (hunk 19 above)
Comment (MaelTeyssedre) [reply ECS/systems/systems.hpp]: non.

File: ECS/systems/systems.hpp (hunk 18 above)
Comment (MaelTeyssedre) [reply ECS/systems/systems.hpp]: non.

File: ECS/components/components.hpp (hunk 17 above)
Comment (MaelTeyssedre) [reply ECS/components/components.hpp:39]: solved

File: ECS/Entity/Entity.hpp (hunk 7 above)
Comment (MaelTeyssedre) [reply ECS/Entity/Entity.hpp:11]: albatar

File: ECS/SparseArray/SparseArray.hpp (hunk 9 above)
Comment (MaelTeyssedre) [reply ECS/SparseArray/SparseArray.hpp:19]: abahui

File: ECS/SparseArray/SparseArray.hpp (hunk 13 above)
Comment (MaelTeyssedre) [reply ECS/SparseArray/SparseArray.hpp:129]: solved

File: ECS/SparseArray/SparseArray.hpp (hunk 15 above)
Comment (MaelTeyssedre) [reply ECS/SparseArray/SparseArray.hpp:316]: solved

**Pull request merged**